lib/internal/streams/web.js

```diff
@@ -0,0 +1,4632 @@
+        signal);
+    } catch (error) {
+      return PromiseReject(error);
+    }
+  }
+
+  /**
+   * @returns {ReadableStream[]}
+   */
+  tee() {
+    if (!isReadableStream(this))
+      throw new ERR_INVALID_THIS('ReadableStream');
+    return readableStreamTee(this, false);
+  }
+
+  /**
+   * @param {{
+   *   preventCancel? : boolean,
+   * }} [options]
+   * @returns {AsyncIterable}
+   */
+  values(options = {}) {
+    if (!isReadableStream(this))
+      throw new ERR_INVALID_THIS('ReadableStream');
+    validateObject(options, 'options');
+    const {
+      preventCancel = false,
+    } = options;
+
+    const reader = new ReadableStreamDefaultReader(this);
+    let done = false;
+    let started = false;
+    let current;
+
+    function nextSteps() {
+      if (done)
+        return PromiseResolve({ done: true, value: undefined });
+
+      if (reader[kState].stream === undefined) {
+        return PromiseReject(
+          new ERR_INVALID_STATE.TypeError(
+            'The reader is not bound to a ReadableStream'));
+      }
+      const promise = createDeferredPromise();
+
+      readableStreamDefaultReaderRead(reader, {
+        [kChunk](chunk) {
+          current = undefined;
+          promise.resolve({ value: chunk, done: false });
+        },
+        [kClose]() {
+          current = undefined;
+          done = true;
+          readableStreamReaderGenericRelease(reader);
+          promise.resolve({ done: true, value: undefined });
+        },
+        [kError](error) {
+          current = undefined;
+          done = true;
+          readableStreamReaderGenericRelease(reader);
+          promise.reject(error);
+        }
+      });
+      return promise.promise;
+    }
+
+    function returnSteps(value) {
+      if (done)
+        return PromiseResolve({ done: true, value });
+      done = true;
+
+      if (reader[kState].stream === undefined) {
+        return PromiseReject(
+          new ERR_INVALID_STATE.TypeError(
+            'The reader is not bound to a ReadableStream'));
+      }
+      assert(!reader[kState].readRequests.length);
+      if (!preventCancel) {
+        const result = readableStreamReaderGenericCancel(reader, value);
+        readableStreamReaderGenericRelease(reader);
+        return PromisePrototypeThen(
+          result,
+          () => { return { done: true, value }; });
+      }
+
+      readableStreamReaderGenericRelease(reader);
+      return PromiseResolve({ done: true, value });
+    }
+
+    // TODO(@jasnell): Explore whether an async generator
+    // can be used here instead of a custom iterator object.
+    return ObjectSetPrototypeOf({
+      next() {
+        // If this is the first read, delay by one microtask
+        // to ensure that the controller has had an opportunity
+        // to properly start and perform the initial pull.
+        // TODO(@jasnell): The spec doesn't call this out so
+        // need to investigate if it's a bug in our impl or
+        // the spec.
```

Comment:
maybe making this `async next() {` with an `await 0;` would give you the microtask more easily?

Reply:
So.. fun story... changing this to an async function that uses an await breaks the WPT's because there is a test for checking that a patched `Promise.prototype.then` is not called and `await ...` uses the `Promise.prototype` ....

![image](https://user-images.githubusercontent.com/439929/123152659-a906d400-d419-11eb-956b-80bf467e00eb.png)

... so while, yes, using await here would be nicer, it's prone to prototype polution. Note... the few places I am using `await` in this code already do not trigger that WPT failure.

Reply:
ah, that is a good point :-/

Reply:
going to leave this as is, without converting to an async function

Reply:
Hmm, that is unfortunate. Using `await` makes the implementation much more readable, but if it's not possible to avoid calling the (potentially patched) global `Promise.prototype.then` then it's a no-go. 😢 

Does other code in Node's `lib` use `await`, and does that code also risk running a potentially monkey-patched `.then`?

> Note... the few places I am using await in this code already do not trigger that WPT failure.

It looks like [the current WPT test with a patched `.then`](https://github.com/web-platform-tests/wpt/blob/master/streams/readable-streams/patched-global.any.js#L140-L141) immediately closes the writable end, and doesn't even wait for the `pipeTo()` promise to settle before completing the test (and restoring the original `.then`). Most likely, if we were to add a test that actually pushes a few chunks through the pipe while `.then` is patched, we'd have more failures. 😕

Reply:
> Does other code in Node's lib use await, and does that code also risk running a potentially monkey-patched .then?

Yes. I'm not sure there's a great way around it other than moving entirely away from using async/await syntax in favor of only using the captured primordials.

Reply:
> > Does other code in Node's lib use await, and does that code also risk running a potentially monkey-patched .then?
> 
> Yes. I'm not sure there's a great way around it other than moving entirely away from using async/await syntax in favor of only using the captured primordials.
> 

One could attach the original `Promise.prototype.then` as an own property and `await` would take it. Hard to say if v8 would recognise to optimise this away though.